src/main/java/com/exadel/discount/service/impl/CouponServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+        log.debug("Getting sorted page-list of all Coupons");
+        List<Coupon> filteredCouponList = couponRepository
+                .findAll(preparePredicateForFindingAllCoupons(couponFilter), paging)
+                .toList();
+        log.debug("Successfully sorted page-list of all Coupons is got");
+        if (filteredCouponList.isEmpty()) {
+            throw new NotFoundException(String.format("No Coupons %s are found", filteredCouponList));
+        }
+        return couponMapper.toCouponDTOList(filteredCouponList);
+    }
+
+    @Override
+    public CouponDTO findCouponById(UUID id) {
+        log.debug("Finding Coupon by ID");
+        Optional<Coupon> couponOptional = couponRepository.findById(id);
+            if (couponOptional.isEmpty()) throw
+                    new NotFoundException(String.format("Coupon with id %s not found", id));
+        log.debug("Successfully Coupon is found by ID");
+        return couponMapper.toCouponDTO(couponOptional.get());
+    }
+
+    public CouponDTO findCouponByDate(LocalDateTime date) {
+        log.debug("Finding coupon by date");
+        Optional<Coupon> coupon = couponRepository.findCouponByDate(date);
+        Coupon foundedCoupon;
+        if (coupon.isPresent()) foundedCoupon = coupon.get();
+        else throw new NotFoundException(String.format("Coupon with date %s not found", date));
+        log.debug("Successfully coupon is found by date");
+        return couponMapper.toCouponDTO(foundedCoupon);
+    }
+
+    @Transactional
+    @Override
+        public CouponDTO assignCouponToUser(CreateCouponDTO createCouponDTO) {
+            log.debug("Finding of certain user by ID");
+
+            User user = userRepository
+                    .findById(createCouponDTO.getUserId())
+                    .orElseThrow(() -> new NotFoundException(String
+                            .format("User with id %s not found", createCouponDTO.getUserId())));
+
+            Discount discount = discountRepository
+                    .findById(createCouponDTO.getDiscountId())
+                    .orElseThrow(() -> new NotFoundException(String
+                            .format("Discount with id %s not found", createCouponDTO.getDiscountId())));
+            log.debug("Successfully certain User and Discount are found by ID. Starting coupon creation/saving.");
+
+            Coupon coupon = new Coupon();
+            coupon.setUser(user);
+            coupon.setDiscount(discount);
+
+            couponRepository.save(coupon);
+            log.debug("Successfully new coupon is saved to certain user");
+            return couponMapper.toCouponDTO(coupon);
+        }
+
+    private Predicate preparePredicateForFindingAllCoupons(CouponFilter couponfilter) {
+        return ExpressionUtils.and(
+                QueryPredicateBuilder.init()
+                        .append(couponfilter.getCountryIds(), QCoupon.coupon.discount.vendorLocations.any()
+                                .city.country.id::in)
```

Comment:
```suggestion
                        .append(couponfilter.getCountryIds(),
                        QCoupon.coupon.discount.vendorLocations.any().city.country.id::in)
```

Reply:
I think it's better to devide two params each on it's own line

Reply:
Got it
Done